### PR TITLE
feat(twofactor): More customization and extended functioniality

### PR DESCRIPTION
**Pull request by TheXorog**
# Description

This touches in the TwoFactor Extension to allow more use-cases, more ways a developer can interact with the result after authentication has taken place and more customization.

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?

Tested with Makoto.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md [hunk 1]
```diff
@@ -88,18 +90,43 @@ To force a command to require two factor, use the [ApplicationCommandRequireEnro
 
 To ask a user to submit their two factor code, use the function [RequestTwoFactorAsync()](xref:DisCatSharp.Extensions.TwoFactorCommands.ApplicationCommands.TwoFactorApplicationCommandExtension#DisCatSharp_Extensions_TwoFactorCommands_ApplicationCommands_TwoFactorApplicationCommandExtension_RequestTwoFactorAsync_BaseContext_) on your [BaseContext](xref:DisCatSharp.ApplicationCommands.Context.BaseContext).
 
-It returns a [TwoFactorResponse](xref:DisCatSharp.Extensions.TwoFactorCommands.Enums.TwoFactorResponse).
-Make sure to annotate your command with the [ApplicationCommandRequireEnrolledTwoFactorAttribute](xref:DisCatSharp.ApplicationCommands.Attributes.ApplicationCommandRequireEnrolledTwoFactorAttribute) attribute.
+It'll return a [TwoFactorResponse](xref:DisCatSharp.Extensions.TwoFactorCommands.Properties.TwoFactorResponse).
```
Comment (Lulalaby) [inline DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md:93]:
```suggestion
It returns a [TwoFactorResponse](xref:DisCatSharp.Extensions.TwoFactorCommands.Properties.TwoFactorResponse).
```

File: DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md [hunk 2]
```diff
@@ -88,18 +90,43 @@ To force a command to require two factor, use the [ApplicationCommandRequireEnro
...
 
 ```cs
 var tfa_result = ctx.RequestTwoFactorAsync();
```
Comment (Lulalaby) [inline DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md:96]:
```suggestion
var tfa_result = await tx.RequestTwoFactorAsync();
```

File: DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md [hunk 3]
```diff
@@ -88,18 +90,43 @@ To force a command to require two factor, use the [ApplicationCommandRequireEnro
...
+Using two factor authentication on buttons is pretty similar to slash commands but it'll need a DiscordClient to attach to.
+
+Run [RequestTwoFactorAsync()](xref:DisCatSharp.Extensions.TwoFactorCommands.ApplicationCommands.TwoFactorApplicationCommandExtension#DisCatSharp_Extensions_TwoFactorCommands_ApplicationCommands_TwoFactorApplicationCommandExtension_RequestTwoFactorAsync_BaseContext_) on your [ComponentInteractionCreateEventArgs](xref:DisCatSharp.EventArgs.ComponentInteractionCreateEventArgs) to ask the user for the two factor auth code.
+
+Same deal as for slash commands, it'll return a [TwoFactorResponse](xref:DisCatSharp.Extensions.TwoFactorCommands.Properties.TwoFactorResponse).
+
+```cs
+async Task SomeButtonInteraction(DiscordClient sender, ComponentInteractionCreateEventArgs e)
+{
+    var tfa_result = e.RequestTwoFactorAsync(sender);
```
Comment (Lulalaby) [inline DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md:118]:
```suggestion
    var tfa_result = await e.RequestTwoFactorAsync(sender);
```

File: DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md [hunk 4]
```diff
@@ -88,18 +90,43 @@ To force a command to require two factor, use the [ApplicationCommandRequireEnro
...
+
+    if (tfa_result != TwoFactorResponse.Result.ValidCode)
```
Comment (Lulalaby) [inline DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md:120]:
```suggestion
    if (tfa_result.Result != TwoFactorResult.ValidCode)
```

File: DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md [hunk 5]
```diff
@@ -88,18 +90,43 @@ To force a command to require two factor, use the [ApplicationCommandRequireEnro
...
 
-It returns a [TwoFactorResponse](xref:DisCatSharp.Extensions.TwoFactorCommands.Enums.TwoFactorResponse).
-Make sure to annotate your command with the [ApplicationCommandRequireEnrolledTwoFactorAttribute](xref:DisCatSharp.ApplicationCommands.Attributes.ApplicationCommandRequireEnrolledTwoFactorAttribute) attribute.
+It'll return a [TwoFactorResponse](xref:DisCatSharp.Extensions.TwoFactorCommands.Properties.TwoFactorResponse).
 
 ```cs
 var tfa_result = ctx.RequestTwoFactorAsync();
 
-if (tfa_result == TwoFactorResponse.ValidCode)
+if (tfa_result != TwoFactorResponse.Result.ValidCode)
```
Comment (Lulalaby) [inline DisCatSharp.Extensions.Docs/articles/extensions/twofactor_commands/intro.md:98]:
```suggestion
if (tfa_result.Result != TwoFactorResult.ValidCode)
```

**Review by Lulalaby — CHANGES_REQUESTED** [review]

File: DisCatSharp.Extensions.TwoFactorCommands/Enums/TwoFactorResponse.cs
Comment (Lulalaby) [inline DisCatSharp.Extensions.TwoFactorCommands/Enums/TwoFactorResponse.cs:1]: rename file

File: DisCatSharp.Extensions.TwoFactorCommands/Properties/TwoFactorResponse.cs
Comment (Lulalaby) [inline DisCatSharp.Extensions.TwoFactorCommands/Properties/TwoFactorResponse.cs:1]: put in folder called Entity
Properties is a reserved folder in c#

**Review by Lulalaby — CHANGES_REQUESTED** [review]

File: DisCatSharp.Extensions.TwoFactorCommands/Enums/TwoFactorResponse.cs
Comment (TheXorog) [reply DisCatSharp.Extensions.TwoFactorCommands/Enums/TwoFactorResponse.cs:1]: I get moving the file but why rename it and to what?

Comment (TheXorog) [reply DisCatSharp.Extensions.TwoFactorCommands/Enums/TwoFactorResponse.cs:1]: nvm

File: DisCatSharp.Extensions.TwoFactorCommands/Entities/TwoFactorResponse.cs [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+
+/// <summary>
+/// The response given upon attempted two factor authentication.
+/// </summary>
+public class TwoFactorResponse
+{
+	/// <summary>
+	/// The modal response <see cref="ComponentInteractionCreateEventArgs"/>. Null if <see cref="Result"/> is <see cref="TwoFactorResult.NotEnrolled"/> or <see cref="TwoFactorResult.TimedOut"/>.
+	/// </summary>
+	public ComponentInteractionCreateEventArgs Interaction { get; internal set; } = null;
```
Comment (Lulalaby) [inline DisCatSharp.Extensions.TwoFactorCommands/Entities/TwoFactorResponse.cs:43]: We use Interaction as name for the entity DiscordInteraction

**Review by Lulalaby — CHANGES_REQUESTED** [review]

File: DisCatSharp.Extensions.TwoFactorCommands/Entities/TwoFactorResponse.cs (hunk 6 above)
Comment (TheXorog) [reply DisCatSharp.Extensions.TwoFactorCommands/Entities/TwoFactorResponse.cs:43]: I know but i wasn't really sure what else to call this, any ideas?

Comment (Lulalaby) [reply DisCatSharp.Extensions.TwoFactorCommands/Entities/TwoFactorResponse.cs:43]: ComponentInteraction maybe

Comment (TheXorog) [reply DisCatSharp.Extensions.TwoFactorCommands/Entities/TwoFactorResponse.cs:43]: Sounds good

**Review by Lulalaby — APPROVED** [review]

**Pull request merged**